File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int maxDia = 0;
        calculateDiameter(root, &maxDia);
        return maxDia;
    }

private:
    int calculateDiameter(TreeNode* root, int* maxDia) {
        if (root == NULL) return 0;

        int leftDia = calculateDiameter(root->left, maxDia);
        int rightDia = calculateDiameter(root->right, maxDia);

        *maxDia = max({*maxDia, leftDia, rightDia, leftDia + rightDia});

        return 1 + max(leftDia, rightDia);
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
    const maxDia = [-1];
    calculateDiameter(root, maxDia);
    return maxDia[0];
};

const calculateDiameter = (root, maxDia) => {
    if (root === null) return 0;

    const leftDia = calculateDiameter(root.left, maxDia);
    const rightDia = calculateDiameter(root.right, maxDia);

    maxDia[0] = Math.max(maxDia, leftDia, rightDia, leftDia + rightDia);

    return 1 + Math.max(leftDia, rightDia);
}